feat: Athena Ambiguity Resolution Algorithm

Ragansu opened the pull request:
This is the PR for Ambiguity Resolution algorithm inspired from ATLAS-Athena

In this PR we are implementing a score based Ambiguity resolution which takes inspiration from ATLAS Ambiguity resolution.

Along with the core Functionality, I have added Algorithm function in Examples, made necessary modification for Pybind and run_full_chain_odd.py and made unit tests.

To replace the detector dependent hard-coding a config file in *.json format can be used. Hence a Json Config Reader is also included in this PR.


This Project was guided by Corentin, Carlo and David.

Regards
Ragansu.

CarloVarni commented:
Thanks @Ragansu I will take a look at this. First comment though is that we need to change the name of the algorithm and remove `Athena` from it

andiwand commented on Examples/Python/tests/test_examples.py:
```diff
@@ -1223,7 +1223,7 @@ def test_ML_Ambiguity_Solver(tmp_path, assert_root_hash):
     try:
         subprocess.check_call(
-            [sys.executable, str(script), "-n1", "--MLSolver"],
+            [sys.executable, str(script), "-n5", "--SolverType ML"],
```
intended change?

Ragansu replied:
-n1 to -n5 was a typo the rest yes.

andiwand commented on Examples/Scripts/Python/full_chain_odd.py:
```diff
@@ -71,10 +73,18 @@
-    "--MLSolver",
-    help="Use the Ml Ambiguity Solver instead of the classical one",
-    action="store_true",
+    "--SolverType",
+    help="Set which ambiguity solver to use, default is the classical one",
+    type=str,
+    default="classical",
```
please use choices https://docs.python.org/3/library/argparse.html#choices

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    // maximum number of shared hit per track
+    std::size_t maxShared = 5;
+
+    double pTMin = 0;
+    double pTMax = 1e9;
+
+    double phiMin = -M_PI;
+    double phiMax = M_PI;
+
+    double etaMin = -5;
+    double etaMax = 5;
```
This seems to be duplicated with `TrackSelector`. Consider reusing the track selector or expecting selected tracks

Ragansu replied:
After discussing with @CarloVarni , Alex and @Corentin-Allaire. we decided the entire scoring could be changed to be less Athena like and to make it more open-ended.  So that probably for  PR.

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    std::vector<std::size_t>& goodTracks) const {
+  auto trackStateContainer = tracks.trackStateContainerHolder();
+  auto trackContainer = std::make_shared<VectorTrackContainer>();
```
I don't think we should assume a backend here. We should rather make this work for all backends or move it out of Core

Ragansu replied:
This was a mistake, the whole function was not needed. The equivalent of this is there in examples. Hence I removed that function entirely

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+  ACTS_VERBOSE("Starting to compute initial state");
+
+  for (const auto& track : tracks) {
+    std::vector<std::tuple<std::size_t, std::size_t, bool>> measurements_tuples;
```
```suggestion
    std::vector<std::tuple<std::size_t, std::size_t, bool>> measurementsTuples;
```

Ragansu replied:
snake case has be replaced with camel case

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+                 detector.otherScoreWeight;
+      }
+
+      // Adding scores based on optional weights
+      for (const auto& ambiweights : optionalCuts.weights) {
+        ambiweights(track, score);
```
```suggestion
      // Adding scores based on optional weights
      for (const auto& ambiweights : optionalCuts.weights) {
        ambiweights(track, score);
```

A bit hard to grasp what is going on from the code itself.
- `optionalCuts` and `weights` (modifier) seems orthogonal
- `ambiweights` does not sound like a function

Ragansu replied:
I have renamed it
```cpp
      for (const auto& weightFunction : optionalCuts.weights) {
        weightFunction(track, score);
      }
```

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+        double p = 1. / log10(10. + 10. * track.chi2() / track.nDoF());
+        if (p > 0) {
+          score += p;
+        } else
+          score -= 50;
```
```suggestion
        } else {
          score -= 50;
        }
```

also: magic number?

Ragansu replied:
I took this numbers from Athena ambiguity Resolution.  But the logic is by default all valid tracks have a starting score of 100 and based on the number of hits,holes etc we will either add or subtract from this number. so 50 is in the same order as 100.  Hence the choice.

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+      // Adding the score based on the chi2/ndf
+      if (track.chi2() > 0 && track.nDoF() > 0) {
+        double p = 1. / log10(10. + 10. * track.chi2() / track.nDoF());
```
```suggestion
        double p = 1. / std::log10(10. + 10. * track.chi2() / track.nDoF());
```

Ragansu replied:
I have made this change

```cpp
        double p = 1. / std::log10(10. + 10. * track.chi2() / track.nDoF());
```

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+      }
+
+      // Adding the score based on the chi2/ndf
+      if (track.chi2() > 0 && track.nDoF() > 0) {
```
beware that `nDoF` is not reduced by the numbers of fitted params

andiwand commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+    if (track.chi2() > 0 && track.nDoF() > 0) {
+      double chi2 = track.chi2();
+      int indf = track.nDoF();
+      double fac = 1. / log10(10. + 10. * chi2 / indf);
```
```suggestion
      double fac = 1. / std::log10(10. + 10. * chi2 / indf);
```

CarloVarni commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+           detectorId++) {
+        auto detector_it = m_cfg.detectorMap.find(detectorId);
+        auto detector = detector_it->second;
+        score += trackFeaturesMap[detectorId].nHits * detector.hitsScoreWeight;
```
What we are doing here is the following: for every track, we run on all the detector Ids. We then check that detector id in the map, if it does not exist we create it (`operator[]`). So basically we have a total of `nTracks * detectorMap.size()` entries.

Is this the expected behaviour?

Ragansu replied:
Yes this is the expected behaviors, The trackFeatureMap stores number of  hits/holes/outliers for  each track per detector, This will substitute and generalize something like npixelhits  or nTRTholes.

CarloVarni replied:
I'm not fully understanding this. We are here considering all detector Ids of the entire detector, right? 
Is the entire detector contributing to the score of a single track?

Corentin-Allaire replied:
What Ragansu means is that the scoring is done separatly for each sub-detector (in ATLAS those would be the Pixel, Strip and TRT). The score weight of a pixel hole is different from a TRT one for example. This also means that almost all of the maps here are 3 or 4 elements map (<10) so them being unsorted won't make any difference

Corentin-Allaire replied:
Reading again your comment, I think the confusing point might be the name detectorID. This is NOT a volume ID, it is an arbitrary id associted with a list of volumes in the volumeMap. For the ODD there are 2 detectorID : 0 and 1 corresponding to pixel and stip

CarloVarni replied:
thanks for the clarification. This is another reason to go for a simpler data structure instead of an `std::map` imho

Corentin-Allaire replied:
I don't fully disagree, a vector would probably be enough. On the other side I am not sure it will make any kind of difference in term of perf between a small map/vector the real question hers is what is the easiest to read/understand/maintain

CarloVarni replied:
I personally find vectors much more understandable. Basically in the code with have M tracks and a N clusters (a maximum of N ... true). So that basically is a matrix. And accessing the elements is a constant operation since we have the indices at hand

Ragansu replied:
I have changed trackFeaturesMap to trackFeaturesVector

CarloVarni commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+  // Loop over all the tracks in the container
+  for (const auto& track : tracks) {
+    auto trackFeaturesMap = std::map<std::size_t, TrackFeatures>();
```
can you remind me why the wave multiple (and independent) maps, one for each track? If these are totally independent, then why can't we process a track at a time and then use the same objects/memory for the other tracks?

Ragansu replied:
The TrackFeatureMap has been changed to TrackFeatureVector

CarloVarni commented on Core/src/AmbiguityResolution/AthenaAmbiguityResolution.cpp:
```diff
@@ -0,0 +1,204 @@
+  std::size_t numberOfTracks = measurementsPerTrack.size();
+  ACTS_DEBUG("Number of tracks: " << numberOfTracks);
+
+  boost::container::flat_map<std::size_t,
```
I would actually suggest a simple `std::vector<int> tracksPerMeasurement (nMeasurements, 0);`

and then, when looping on the tracks you simply increment the corresponding element (i.e. `++tracksPerMeasurement[iMeasurement];`)

In the end you simply check the size of the sets. So an integer will do the same

Ragansu replied:
I took this implementation from greedy solver.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+/// 4) Remove tracks that have a score below a certain threshold or not have
+/// enough hits 5) Remove tracks that are not good enough based on cuts Contains
+/// method for data preparations
+class AthenaAmbiguityResolution {
```
What about something like ScoreBasedAmbiguityResolution ?

CarloVarni replied:
Yes, that is better indeed

Ragansu replied:
I have changed the names everywhere from AthenaAmbiguityResolution to ScoreBasedAmbiguityResolution

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    // maximum number of shared hit per track
+    std::size_t maxShared = 5;
+
+    double pTMin = 0;
+    double pTMax = 1e9;
```
Yes, I will add units to them.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    std::size_t maxOutliers;
+    std::size_t maxSharedHits;
+
+    bool sharedHitsFlag;  // if true, the shared hits are considered as bad hits
+                          // for this detector
+
+    std::size_t detectorId;
+
+    std::vector<double>
+        factorHits;  // a list of values from  0 to 1, the higher number of
+                     // hits, higher value in the list is multiplied to
+                     // ambuiguity score
+                     // applied only if useAmbiguityFunction is true
+    std::vector<double>
+        factorHoles;  // a list of values from  0 to 1, the higher number of
+                      // holes, lower value in the list is multiplied to
+                      // ambuiguity score
+                      // applied only if useAmbiguityFunction is true
```
I will redo the comments to work with doxygen.

Ragansu commented:
I have replied to some comments which I have fully understood. The rest especially things connected with templates and std::functions, I will discuss with @Corentin-Allaire  and @CarloVarni  and reply. Thank You all for the comments.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+        if (volume_it == m_cfg.volumeMap.end()) {
+          ACTS_ERROR("Volume " << iVolume << "not found in the volume map");
+          continue;
+        }
```
I did this. It was originally like this is itself but I ran into some segfaults while applied in athena.

Ragansu replied:
I have now merged this part into compute initial state there by moving it out off scoring
```cpp

    for (auto ts : track.trackStatesReversed()) {
      auto* referenceSurfacePtr = &ts.referenceSurface();
      if (referenceSurfacePtr == nullptr) {
        ACTS_ERROR("Track state has no reference surface");
        continue;
      }
      auto iVolume = ts.referenceSurface().geometryId().volume();
      auto volume_it = m_cfg.volumeMap.find(iVolume);
      if (volume_it == m_cfg.volumeMap.end()) {
        ACTS_ERROR("Volume " << iVolume << "not found in the volume map");
        continue;
      }
      auto detectorId = volume_it->second;

      if (ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
        Acts::SourceLink sourceLink = ts.getUncalibratedSourceLink();
        ACTS_DEBUG("Track state type is MeasurementFlag");

        if (ts.typeFlags().test(Acts::TrackStateFlag::SharedHitFlag)) {
          trackFeaturesVector[detectorId].nSharedHits++;
        }
        trackFeaturesVector[detectorId].nHits++;

        // assign a new measurement index if the source link was not seen yet
        auto emplace = measurementIndexMap.try_emplace(
            sourceLink, measurementIndexMap.size());

        bool isoutliner = false;

        measurementTuples.push_back(
            std::make_tuple(emplace.first->second, iVolume, isoutliner));
      } else if (ts.typeFlags().test(Acts::TrackStateFlag::OutlierFlag)) {
        Acts::SourceLink sourceLink = ts.getUncalibratedSourceLink();
        ACTS_DEBUG("Track state type is OutlierFlag");
        trackFeaturesVector[detectorId].nOutliers++;

        // assign a new measurement index if the source link was not seen yet
        auto emplace = measurementIndexMap.try_emplace(
            sourceLink, measurementIndexMap.size());

        bool isoutliner = true;

        measurementTuples.push_back(
            std::make_tuple(emplace.first->second, iVolume, isoutliner));
      } else if (ts.typeFlags().test(Acts::TrackStateFlag::HoleFlag)) {
        ACTS_DEBUG("Track state type is HoleFlag");
        trackFeaturesVector[detectorId].nHoles++;
      }
    }
    measurementsPerTrack.push_back(std::move(measurementTuples));
    trackFeaturesVectors.push_back(std::move(trackFeaturesVector));
  }
```

Ragansu commented on Core/src/AmbiguityResolution/AthenaAmbiguityResolution.cpp:
```diff
@@ -0,0 +1,204 @@
+  std::size_t numberOfTracks = measurementsPerTrack.size();
+  ACTS_DEBUG("Number of tracks: " << numberOfTracks);
+
+  boost::container::flat_map<std::size_t,
```
I took it from greedy solver.

Ragansu commented:
I have made some changes based on the comments and pushed a new commit to this PR. I have resolved the comments which I have addressed.  Rest of the comments I will address after the discussion with @CarloVarni , @Corentin-Allaire  and 	
Alexis Vallier on Monday.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    double pTMax = 1e9;
+
+    double phiMin = -M_PI;
+    double phiMax = M_PI;
```
I have added units

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  /// The configuration can be saved in a json file and loaded from there.
+  ///
+  struct DetectorConfig {
+    int hitsScoreWeight;
```
I have initialized the values to 0

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  ///
+  /// The trackFeatures is used to compute the score of each track
+  struct TrackFeatures {
+    std::size_t nHits;
```
I have initialized the values to 0

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+
+    bool useAmbiguityFunction =
+        false;  // if true, the ambiguity score is
+                // computed based on a different function.
```
I have made that change

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    const Acts::TrackProxy<track_container_t, trajectory_t, holder_t, true>&,
+    double&)>;
+
+}  // namespace AthenaAmbiguitySolver
```
I have moved the OptionalFilter to the struct OptionalCuts which is in ScoreBasedAmbiguityResolution (new name of AthenaAmbiguityResolution)

Ragansu replied:
```cpp
  struct Optional_cuts {
    using OptionalFilter =
        std::function<bool(const Acts::TrackProxy<track_container_t, traj_t,
                                                  holder_t, ReadOnly>&)>;

    using OptionalScoreModifier = std::function<void(
        const Acts::TrackProxy<track_container_t, traj_t, holder_t, ReadOnly>&,
        double&)>;
    std::vector<OptionalFilter> cuts = {};
    std::vector<OptionalScoreModifier> weights = {};

    /// applied only if useAmbiguityFunction is true
    std::vector<OptionalScoreModifier> ambiscores = {};
  };
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+            template <typename> class holder_t>
+  std::vector<int> solveAmbiguity(
+      const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+      std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
```
I have defined a typedef called measurementTuple for this `  using measurementTuple = std::tuple<std::size_t, std::size_t, bool>;
`

Ragansu commented on Plugins/Json/include/Acts/Plugins/Json/AmbiguityConfigJsonConverter.hpp:
```diff
@@ -0,0 +1,34 @@
+#include "Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp"
+#include "Acts/Plugins/Json/ActsJson.hpp"
+
+#include <fstream>
```
I have removed the `#include <fstream>`

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    const Acts::TrackProxy<track_container_t, trajectory_t, holder_t, true>&,
+    double&)>;
+
+}  // namespace AthenaAmbiguitySolver
```
I have made this change
```cpp
  struct Optional_cuts {
    using OptionalFilter =
        std::function<bool(const Acts::TrackProxy<track_container_t, traj_t,
                                                  holder_t, ReadOnly>&)>;

    using OptionalScoreModifier = std::function<void(
        const Acts::TrackProxy<track_container_t, traj_t, holder_t, ReadOnly>&,
        double&)>;
    std::vector<OptionalFilter> cuts = {};
    std::vector<OptionalScoreModifier> weights = {};

    /// applied only if useAmbiguityFunction is true
    std::vector<OptionalScoreModifier> scores = {};
  };
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+template <typename track_container_t, typename trajectory_t,
+          template <typename> class holder_t>
+using OptinalFilter = std::function<bool(
+    const Acts::TrackProxy<track_container_t, trajectory_t, holder_t, true>&)>;
```
I have templated this part
```cpp
  template <typename track_container_t, typename traj_t,
            template <typename> class holder_t, bool ReadOnly>
  struct Optional_cuts {
    using OptionalFilter =
        std::function<bool(const Acts::TrackProxy<track_container_t, traj_t,
                                                  holder_t, ReadOnly>&)>;
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+        factorHoles;  // a list of values from  0 to 1, the higher number of
+                      // holes, lower value in the list is multiplied to
+                      // ambuiguity score
+                      // applied only if useAmbiguityFunction is true
```
I have added initial value

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+    std::size_t nHoles;
+    std::size_t nOutliers;
+    std::size_t nSharedHits;
+  };
```
I have initialized all the members to 0

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  /// @brief Configuration struct : contains the configuration for the ambiguity resolution.
+  struct Config {
+    std::map<std::size_t, std::size_t> volumeMap = {{0, 0}};
+    std::map<std::size_t, DetectorConfig> detectorMap;
```
This has been changed to vector 
```cpp
    std::vector<TrackFeatures> trackFeaturesVector(numberOfDetectors);
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  template <typename track_container_t, typename traj_t,
+            template <typename> class holder_t>
+  struct Optional_cuts {
+    std::vector<AthenaAmbiguitySolver::OptinalFilter<track_container_t, traj_t,
```
As we discussed on Monday we could do it in another PR along with other changes like making scoring completely open ended.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  template <typename track_container_t, typename traj_t,
+            template <typename> class holder_t>
+  struct Optional_cuts {
+    std::vector<AthenaAmbiguitySolver::OptinalFilter<track_container_t, traj_t,
```
Typo fixed

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  std::vector<double> simpleScore(
+      const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+      std::vector<std::map<std::size_t, TrackFeatures>>& trackFeaturesMaps,
+      Optional_cuts<track_container_t, traj_t, holder_t> optionalCuts = {})
```
I have added constness to where ever it made sense. 

```cpp
template <typename track_container_t, typename traj_t,
          template <typename> class holder_t, bool ReadOnly>
std::vector<double> Acts::ScoreBasedAmbiguityResolution::simpleScore(
    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
    const std::vector<std::vector<TrackFeatures>>& trackFeaturesVectors,
    const OptionalCuts<track_container_t, traj_t, holder_t, ReadOnly>&
        optionalCuts) const {
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  /// @param measurementsPerTrack is the list of measurements for each track
+  /// @return a vector of IDs of the tracks we want to keep
+  std::vector<std::size_t> getCleanedOutTracks(
+      std::vector<double> trackScore,
```
I have added constness and reference here. 
```cpp
  std::vector<bool> getCleanedOutTracks(
      const std::vector<double>& trackScore,
      const std::vector<std::vector<TrackFeatures>>& trackFeaturesVectors,
      const std::vector<std::vector<measurementTuple>>& measurementsPerTrack)
      const;
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  std::vector<std::size_t> getCleanedOutTracks(
+      std::vector<double> trackScore,
+      std::vector<std::map<std::size_t, TrackFeatures>>& trackFeaturesMaps,
+      std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
```
I have added const reference for all the parameters in this function.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  Config m_cfg;
+
+  /// Logging instance
+  std::unique_ptr<const Logger> m_logger;
```
I have made this change

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  std::unique_ptr<const Logger> m_logger;
+
+  /// Private access to logging instance
+  const Logger& logger() const { return *m_logger; }
```
I have moved this to `ipp`
```cpp
const Logger& ScoreBasedAmbiguityResolution::logger() const {
  return *m_logger;
}
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+    for (auto ts : track.trackStatesReversed()) {
+      ACTS_DEBUG("Track state type check: ");
+      if (ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
```
I have redone the `computeinitialstate()`

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+                         source_link_equality_t>(0, sourceLinkHash,
+                                                 sourceLinkEquality);
+
+  std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
```
I have done this
```cpp
  std::vector<std::vector<measurementTuple>> measurementsPerTrack;
  measurementsPerTrack.reserve(tracks.size());
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+  ACTS_VERBOSE("Starting to compute initial state");
+
+  for (const auto& track : tracks) {
+    std::vector<std::tuple<std::size_t, std::size_t, bool>> measurements_tuples;
```
I tried that with 
```cpp
    measurementTuples.reserve(track.trackSateReserved().size());
```
But was unsuccessful. I will try to fix this.

Ragansu replied:
I found a solution with std::distance

```cpp
    int numberOfTrackStates = std::distance(track.trackStatesReversed().begin(),
                                            track.trackStatesReversed().end());
    std::vector<measurementTuple> measurementTuples;
    measurementTuples.reserve(numberOfTrackStates);
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+  // Loop over all the tracks in the container
+  for (const auto& track : tracks) {
+    auto trackFeaturesMap = std::map<std::size_t, TrackFeatures>();
```
I have replaced it with vectors
```cpp

    const auto trackFeaturesVector =
        trackFeaturesVectors[iTrack]; 
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+          template <typename> class holder_t>
+std::vector<double> Acts::AthenaAmbiguityResolution::simpleScore(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    std::vector<std::map<std::size_t, TrackFeatures>>& trackFeaturesMaps,
```
Its now replaced with vector of vectors
```cpp
std::vector<double> Acts::ScoreBasedAmbiguityResolution::simpleScore(
    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
    const std::vector<std::vector<TrackFeatures>>& trackFeaturesVectors,
    const OptionalCuts<track_container_t, traj_t, holder_t, ReadOnly>&
        optionalCuts) const {
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+    double score = 1;
+    // cuts on pT
+    if (Acts::VectorHelpers::perp(track.momentum()) > m_cfg.pTMax ||
```
I have redone it 
```cpp
    auto pT = Acts::VectorHelpers::perp(track.momentum());
    auto eta = Acts::VectorHelpers::eta(track.momentum());
    auto phi = Acts::VectorHelpers::phi(track.momentum());
    // cuts on pT
    if (pT < m_cfg.pTMin || pT > m_cfg.pTMax) {
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+
+      ACTS_DEBUG("---> Found summary information");
+      ACTS_DEBUG("---> Detector ID: " << detectorId);
+      ACTS_DEBUG("---> Number of hits: " << trackFeaturesMap[detectorId].nHits);
```
now that the trackfeatureMap is replaces by trackFeatureVector this is lo longer an issue

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+    iTrack++;
+
+    // Add the score to the vector
+    trackScore.push_back(score);
```
the useAmbiguityFunction execution has been simplified.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+  TrackContainer solvedTracks{trackContainer, tempTrackStateContainer};
+  solvedTracks.ensureDynamicColumns(tracks);
+
+  for (auto&& iTrack : goodTracks) {
```
This is removed as we redid good tracks

Ragansu commented on Core/src/AmbiguityResolution/AthenaAmbiguityResolution.cpp:
```diff
@@ -0,0 +1,204 @@
+#include <stdexcept>
+
+std::vector<std::size_t> Acts::AthenaAmbiguityResolution::getCleanedOutTracks(
+    std::vector<double> trackScore,
```
I have redone it with  const reference

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  computeInitialState(
+      const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+      source_link_hash_t&& sourceLinkHash,
+      source_link_equality_t&& sourceLinkEquality) const;
```
I have taken this straight from greedy solver I am not very sure about it.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,223 @@
+  ///
+  /// The configuration can be saved in a json file and loaded from there.
+  ///
+  struct DetectorConfig {
```
These are detector specific cuts and scores for the ambiguity resolution, so it makes sense to add it in the class definition. but maybe in the future when we address the scoring again we can think about separating it into another files.

Ragansu commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+  std::vector<int> goodTracks;
+  std::size_t iTrack = 0;
+  for (const auto& track : tracks) {
+    if (std::find(cleanTracks.begin(), cleanTracks.end(), iTrack) !=
```
I have modified the clean trcaks to vector of bools
```cpp
  for (const auto& track : tracks) {
    if (cleanTracks[iTrack]) {
      cleanTrackIndex++;
      if (trackScore[iTrack] >= m_cfg.minScore) {
        goodTracks.push_back(track.index());
      }
    }
    iTrack++;
  }
```

Ragansu commented:
I have made the necessary modification. except in places concerning source-link or boost containers as I took those from greedy solver and I am not sure exactly how to address thoes questions.

As for the rest, I have replies with snippets of code  for all the major comments. 


@CarloVarni , @andiwand and @Corentin-Allaire  Please review this. 

Thank You

CarloVarni commented on Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,406 @@
+  auto measurementIndexMap =
+      std::unordered_map<SourceLink, std::size_t, source_link_hash_t,
+                         source_link_equality_t>(0, sourceLinkHash,
+                                                 sourceLinkEquality);
```
this is ok for the time being

CarloVarni commented on Examples/Algorithms/AmbiguityResolution/src/AthenaAmbiguityResolutionAlgorithm.cpp:
```diff
@@ -0,0 +1,139 @@
+std::size_t sourceLinkHash(const Acts::SourceLink& a) {
+  return static_cast<std::size_t>(
+      a.get<ActsExamples::IndexSourceLink>().index());
+}
```
this will need some thinking. Let's keep it as it is for the time being

andiwand replied:
I think it would be good to elevate this into Core but I agree that we can do this later

CarloVarni commented on Core/src/AmbiguityResolution/ScoreBasedAmbiguityResolution.cpp:
```diff
@@ -0,0 +1,204 @@
+    // trackStateTypes and other conditions.
+    // Good measurements are copied to the newMeasurementsPerTrack vector.
+    for (std::size_t i = 0; i < trackStateTypes.size(); i++) {
+      auto measurementTuples = measurementsPerTrack[iTrack][i];
```
```suggestion
     auto& measurementTuples = measurementsPerTrack[iTrack][i];
```

CarloVarni reviewed (commented):
Some additional comments. Please take a look and check for possible unwanted copies.

andiwand commented:
One general comment about the name `ScoreBasedAmbiguityResolution`:

I think it is much better than than `AtlasAmbiguityResolution` but I am not sure if it captures the difference to the `GreedyAmbiguityResolution`. The greedy one is also doing scoring in terms of numbers of measurements and chi2 and this is also something I imagined to add a customization point in the future.

Do you see a way to capture some essential difference between the two algorithms in the name?

If not I wonder if just the way of scoring is difference and the behavior of the Atlas one can be achieved by customizing the greedy one.

Ragansu commented:
> One general comment about the name `ScoreBasedAmbiguityResolution`:
> 
> I think it is much better than than `AtlasAmbiguityResolution` but I am not sure if it captures the difference to the `GreedyAmbiguityResolution`. The greedy one is also doing scoring in terms of numbers of measurements and chi2 and this is also something I imagined to add a customization point in the future.
> 
> Do you see a way to capture some essential difference between the two algorithms in the name?
> 
> If not I wonder if just the way of scoring is difference and the behavior of the Atlas one can be achieved by customizing the greedy one.

So the main difference between greedy and ScoreBased solver is in the way it handles shared hits. In greedy it used the chi2 or number of hits to decide between 2 tracks.  but ScoreBased or AthenaSolver doesnt compare tracks directly instead it removed shared hits if the tracks has a lower score. and the tracks with not enough hits is removed later.

Ragansu commented:
I rechecked for any unwanted coping and I replaced few initialization with const ref. 
I also noticed there was a map which could be replaced with a vector and I replaced it.

andiwand commented on Plugins/Json/include/Acts/Plugins/Json/AmbiguityConfigJsonConverter.hpp:
```diff
@@ -0,0 +1,31 @@
+
+namespace Acts {
+
+/// @class ScoreBasedAmbiguityResolutionJsonConverter
+///
+/// @brief A converter class for the ScoreBasedAmbiguityResolution algorithm
+///
+/// This class is used to convert the ScoreBasedAmbiguityResolutionCongig
+/// from JSON to C++
+class AmbiguityConfigJsonConverter {
+  using DetectorConfig = ScoreBasedAmbiguityResolution::DetectorConfig;
+
+ public:
+  std::pair<std::map<std::size_t, std::size_t>,
+            std::vector<ScoreBasedAmbiguityResolution::DetectorConfig>>
+  fromJson(const std::string&) const;
+};
```
I don't think this needs to be a class? Usually we just overload `from_json`

Ragansu replied:
Done

andiwand commented on Tests/UnitTests/Core/AmbiguityResolution/ScoreBasedAmbiguityResolutionTest.cpp:
```diff
@@ -0,0 +1,132 @@
+
+using Acts::MultiTrajectoryTraits::IndexType;
+
+namespace Acts {
+
+namespace Test {
```
```suggestion
namespace Acts::Test {
```

Ragansu commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,358 @@
+
+namespace Acts {
+
+inline const Logger& ScoreBasedAmbiguityResolution::logger() const {
+  return *m_logger;
+}
```
you mean in hpp ?

andiwand replied:
right

Ragansu replied:
@CarloVarni  was in the opinion its better in ipp since its inline.

CarloVarni replied:
I indeed prefer implementation in `ipp` and declarations in `hpp` so that you have all implementations in one single place

andiwand replied:
makes sense 👍 it is just a personal preference to put one liners into the header. no need to change it

Ragansu commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,358 @@
+
+template <typename track_container_t, typename traj_t,
+          template <typename> class holder_t, bool ReadOnly>
+std::vector<double> Acts::ScoreBasedAmbiguityResolution::simpleScore(
```
I can split it into 2 call one or the other from solveambiguity

Ragansu replied:
I have split into simpleScore and AmbiguityScore

Ragansu commented on Examples/Python/python/acts/examples/reconstruction.py:
yes. It did pass the format_py check in acts (push)

andiwand replied:
I does pass but I would still prefer sticking to the old format as it seems more concise and also passes the check

Ragansu replied:
I redid the formatting this time with black directly instead of using VS extension. Hope its better now.

andiwand replied:
There are still some unrelated formatting changes. I would prefer to revert them but that is just a personal preference

Ragansu commented on Examples/Scripts/Python/MLAmbiguityResolution/seedDuplicateClassifier.onnx:
its an issue in seeding and not in Ambiguity, The file name was different, so I renamed it to run the odd with ml seeding. I think that ended up duplicating the file instead of renaming. I could remove it.

@Corentin-Allaire  should I remove it in this PR ?

Corentin-Allaire replied:
It is just a renaming write ? Might as well do this here we could one a PR just for that but...

andiwand replied:
But from GitHub it looks like you are adding a file with 100kB not just renaming it. To make the rename effective you need to delete the old file.

Corentin-Allaire replied:
Ah yes ! Good point, I didn't check this. (ideally you should use git mv ... but it is probably too late)

Ragansu replied:
I have removed the old file.

Ragansu commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,205 @@
+/// data preparations
+class ScoreBasedAmbiguityResolution {
+ public:
+  using measurementTuple = std::tuple<std::size_t, std::size_t, bool>;
```
I have changed them to structs

Ragansu commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,205 @@
+  /// The optional cuts,weights and score are used to remove tracks that are not
+  /// good enough, based on some criteria. Users are free to add their own cuts
+  /// with the help of this struct.
+  template <typename track_container_t, typename traj_t,
+            template <typename> class holder_t, bool ReadOnly>
+  struct OptionalCuts {
+    using OptionalFilter =
+        std::function<bool(const Acts::TrackProxy<track_container_t, traj_t,
+                                                  holder_t, ReadOnly>&)>;
```
I didn't quiet understand what you meant.   you mean we could directly use `track_container_t::TrackProxy` instead of Acts::TrackProxy ?

andiwand replied:
Instead of templating `template <typename track_container_t, typename traj_t, template <typename> class holder_t, bool ReadOnly>` everywhere you could also just use `template <typename track_container_t>` and the user can point it to `Acts::TrackContainer`. I think that does the same but replaces the various template args with a single one with is the one you use at the end.

It's not so important and can also be followed up later

Ragansu replied:
I think I will maybe do it in a different PR as the templating part is bound to change when we change scoring and optional cuts.

CarloVarni commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,469 @@
+std::vector<std::vector<ScoreBasedAmbiguityResolution::MeasurementInfo>>
+ScoreBasedAmbiguityResolution::computeInitialState(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    source_link_hash_t&& sourceLinkHash,
+    source_link_equality_t&& sourceLinkEquality,
```
here too

CarloVarni commented:
One last comment. Then there is only the open thread from @andiwand to be addressed.

andiwand commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:
```diff
@@ -0,0 +1,469 @@
+        auto emplace = MeasurementIndexMap.try_emplace(
+            sourceLink, MeasurementIndexMap.size());
+
+        bool isoutliner = true;
```
```suggestion
        bool isOutliner = true;
```

andiwand commented on Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:
```diff
@@ -0,0 +1,231 @@
+
+  /// @brief MeasurementInfo : contains the measurement ID and the detector ID
+  struct MeasurementInfo {
+    std::size_t iMeasurement;
+    std::size_t detectorId;
+    bool isOutlier;
```
```suggestion
    std::size_t iMeasurement = 0;
    std::size_t detectorId = 0;
    bool isOutlier = false;
```

andiwand commented on Plugins/Json/src/AmbiguityConfigJsonConverter.cpp:
```diff
@@ -0,0 +1,76 @@
+
+std::pair<std::map<std::size_t, std::size_t>,
+          std::vector<ScoreBasedAmbiguityResolution::DetectorConfig>>
+from_json(const std::string& configFile) {
+  std::ifstream file(configFile);
+  if (!file.is_open()) {
+    std::cerr << "Error opening file: " << configFile << std::endl;
+    return {};
+  }
+
+  std::cout << "Reading configuration file: " << configFile << std::endl;
+
+  nlohmann::json j;
+
+  file >> j;
```
Usually this function would take a `json` object and then turns it into something else. you can find an example here https://github.com/acts-project/acts/blob/bed53b2fd3ddffc36cc7c8176ea4bbb447e6ec2e/Plugins/Json/include/Acts/Plugins/Json/MaterialJsonConverter.hpp#L30

Ragansu replied:
Okay I will do it like that.

Ragansu replied:
I have changed the json to this style

andiwand commented on Examples/Scripts/Python/full_chain_odd.py:
There are still some unrelated formatting changes. I would prefer to revert them but that is just a personal preference

andiwand commented on Examples/Python/tests/test_examples.py:
```diff
@@ -1223,7 +1223,7 @@ def test_ML_Ambiguity_Solver(tmp_path, assert_root_hash):
     try:
         subprocess.check_call(
-            [sys.executable, str(script), "-n1", "--MLSolver"],
+            [sys.executable, str(script), "-n1", "--SolverType ML"],
```
this needs to be updated after the flag changed

Ragansu replied:
I have updated the flag and error has been fixed. The test is not failing anymore.

andiwand reviewed (commented):
We are pretty close now. I left a last round of comments. @Corentin-Allaire @CarloVarni feel free to go ahead with the merge

CarloVarni commented:
there is a failing test with `FAILED Examples/Python/tests/test_examples.py::test_ML_Ambiguity_Solver - AttributeError: 'NoneType' object has no attribute 'decode'` and the format to fix

andiwand commented:
Right I think this is because of the changed flag? Also the formatting job is not happy at the moment

Ragansu commented:
I have fixed the issue with the test_examples.py and the code passed all build tests in my fork.
![image](https://github.com/acts-project/acts/assets/66349236/5619e031-9678-4447-a1d7-e54ebf227a44)

andiwand commented:
@paulgessinger could you add @Ragansu to the CI permission groups?

paulgessinger commented:
@andiwand Invite sent

Ragansu commented:
Thank You @paulgessinger  I got the invite.

CarloVarni commented:
I'm good with this. @andiwand do you have additional comments?

CarloVarni commented:
approving from my side

andiwand commented:
There are a few clang-tidy failures which require fixing https://github.com/acts-project/acts/pull/3139/checks?check_run_id=24637549405

Ragansu commented:
> That is a big junk of work - Thank you @Ragansu !

Thank you so much for this. I learned a lot during this exercise.

acts-project-service commented:
## :red_circle: Athena integration test results [91e7033528d19d4258e5b9709fd8f750f8a608f9]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7347592)!

|status|job|report|
|------|---|------|
|:red_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38711670)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707352)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707351)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707350)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707349)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707348)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707347)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707346)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707345)||
|:red_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707344)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707343)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707342)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707341)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707340)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707339)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707338)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707337)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707336)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707335)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707334)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707333)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707332)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707331)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707330)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707329)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707328)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707327)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707326)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707325)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38707324)||
